File: Solutions/FloodFill.cpp
// Leetcode #733: https://leetcode.com/problems/flood-fill/

class Solution {
public:
    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int color) {
        // If the starting pixel needs to be changed, call the helper function
        if (image[sr][sc] != color) helper(image, sr, sc, color, image[sr][sc]);
        
        return image; // Return the vector
    }
    
    void helper(vector<vector<int>>& image, int sr, int sc, int color, int match) {
        // If the pixel does not match the color of the starting pixel, just return
        if (image[sr][sc] != match) return;
        
        // Change the color of the current pixel
        image[sr][sc] = color;
        
        // Recursively flood fill any adjacent pixels
        if (sr - 1 >= 0) helper(image, sr - 1, sc, color, match);                   // image[sr - 1][sc]
        if (sr + 1 < image.size()) helper(image, sr + 1, sc, color, match);         // image[sr + 1][sc]
        if (sc - 1 >= 0) helper(image, sr, sc - 1, color, match);                   // image[sr][sc - 1]
        if (sc + 1 < image[sr].size()) helper(image, sr, sc + 1, color, match);     // image[sr][sc + 1]
    }
};

File: Solutions/FloodFill.py
# Leetcode #733: https://leetcode.com/problems/flood-fill/description/
# Source: https://leetcode.com/problems/flood-fill/solutions/109604/easy-python-dfs-no-need-for-visited/?orderBy=most_votes

# Solution 1 (Recursive DFS) | O(n) time, O(n) space
class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
        # Helper function that recursively flood fills using dfs
        def dfs(row, col):
            # Check if row and col are valid, and if its color is same as starting cell's color
            if (0 <= row < rows and 0 <= col < cols) and image[row][col] == startColor:
                # Change the current cell's color
                image[row][col] = color

                # Recursively flood fill its neighboring cells
                # [dfs(row + x, col + y) for (x, y) in ((0, 1), (1, 0), (0, -1), (-1, 0))]  # alternative to code below
                dfs(row - 1, col)
                dfs(row + 1, col)
                dfs(row, col - 1)
                dfs(row, col + 1)

        rows, cols, startColor = len(image), len(image[0]), image[sr][sc]

        # If the starting pixel's color isn't the same as target color, call the helper function
        if startColor != color:
            dfs(sr, sc)

        return image
